segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,263 @@
+     * If suspended or there are no items then it "sleeps" for time specified by configuration.
+     *
+     * @param isBackground True if the caller is backgound task else False if called explicitly.
+     * @param maxItems Maximum number of items to delete at a time.
+     * @return CompletableFuture which is completed when garbage is deleted.
+     */
+    CompletableFuture<Void> deleteGarbage(boolean isBackground, int maxItems) {
+        // Sleep if suspended.
+        if (suspended.get() && isBackground) {
+            log.info("{} deleteGarbage - suspended - sleeping for {}.", chunkedSegmentStorage.getLogPrefix(), config.getGarbageCollectionDelay());
+            return Futures.delayedFuture(config.getGarbageCollectionSleep(), executor);
+        }
+
+        // Find chunks to delete.
+        val currentTime = System.currentTimeMillis();
+        val chunksToDelete = new ArrayList<GarbageChunkInfo>();
+        int count = 0;
+        val iterator = garbageChunks.iterator();
+        while (iterator.hasNext()) {
+            GarbageChunkInfo info = iterator.next();
+            if (info == null || count >= maxItems) {
+                break;
+            }
+            if (canDelete(info.getDeletedTime(), currentTime)) {
+                queueSize.decrementAndGet();
+                chunksToDelete.add(info);
+                iterator.remove();
+                count++;
+            }
+        }
+
+        // Sleep if no chunks to delete.
+        if (count == 0) {
+            log.debug("{} deleteGarbage - no work - sleeping for {}.", chunkedSegmentStorage.getLogPrefix(), config.getGarbageCollectionDelay());
+            return Futures.delayedFuture(config.getGarbageCollectionSleep(), executor);
+        }
+
+        // For each chunk delete if the chunk is not present at all in the metadata or is present but marked as inactive.
+        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (val infoToDelete : chunksToDelete) {
+            val chunkToDelete = infoToDelete.name;
+            val txn = chunkedSegmentStorage.getMetadataStore().beginTransaction(false, chunkToDelete);
+            val future =
+                    txn.get(infoToDelete.name)
+                    .thenComposeAsync( metadata -> {
+                            val chunkMetadata = (ChunkMetadata) metadata;
+                            // Delete if the chunk is not present at all in the metadata or is present but marked as inactive.
+                            boolean shouldDelete = (null == chunkMetadata) ? true : !chunkMetadata.isActive();
+                            // Check whether the chunk is marked as inactive
+                            if (shouldDelete) {
+                                return chunkedSegmentStorage.getChunkStorage().openWrite(chunkToDelete)
+                                        .thenComposeAsync(chunkedSegmentStorage.getChunkStorage()::delete, executor)
```

Comment:
We discussed that this could lead to another problem: if the chunk is deleted and then something fails, the metadata will continue reflecting that the chunk is there and needs deletion. The same chunk will be picked, which has been deleted from storage, and the deletion will fail, skipping again the deletion in metadata. So this could lead to an infinite loop. Please, either modify this logic to handle the metadata deletion first (so in the worst case if there is a failure after that we would leave one orphan chunk to delete), or handle the case in which the chunk deletion succeeds bu there is an exception just after that, so we can eventually delete it from metadata too and avoid the infinite retries on that chunk.

Reply:
Also, the fact that we haven't found this problem in the test cases of this PR makes me feel that more effort is necessary on testing the potential scenarios that could happen to the garbage collector (fail in chunk deletion but not in metadata, the other way raound, etc.).

Reply:
added tests